File: slcm/admission/report/applicant_fee_payments/applicant_fee_payments.py
import frappe
from frappe import _

def execute(filters: dict | None = None):
	columns = get_columns()
	data = get_data(filters)
	
	# Returning column and data, other parts handled via JS
	return columns, data, None, None, None

def get_columns() -> list[dict]:
	return [
		{
			"label": _("Receipt ID"),
			"fieldname": "name",
			"fieldtype": "Link",
			"options": "Applicant Payment Receipt",
			"width": 140
		},
		{
			"label": _("Applicant"),
			"fieldname": "applicant",
			"fieldtype": "Link",
			"options": "Applicant",
			"width": 120
		},
		{
			"label": _("Applicant Name"),
			"fieldname": "applicant_name",
			"fieldtype": "Data",
			"width": 180
		},
		{
			"label": _("Program"),
			"fieldname": "program",
			"fieldtype": "Link",
			"options": "Program",
			"width": 180
		},
		{
			"label": _("Campus"),
			"fieldname": "campus",
			"fieldtype": "Link",
			"options": "Campus",
			"width": 120
		},
		{
			"label": _("Payment Date"),
			"fieldname": "payment_date",
			"fieldtype": "Date",
			"width": 120
		},
		{
			"label": _("Payment Mode"),
			"fieldname": "payment_mode",
			"fieldtype": "Data",
			"width": 120
		},
		{
			"label": _("Transaction ID"),
			"fieldname": "transaction_id",
			"fieldtype": "Data",
			"width": 140
		},
		{
			"label": _("Amount"),
			"fieldname": "total_amount",
			"fieldtype": "Currency",
			"options": "currency",
			"width": 120
		},
		{
			"label": _("Currency"),
			"fieldname": "currency",
			"fieldtype": "Link",
			"options": "Currency",
			"hidden": 1
		}
	]

def get_data(filters: dict | None) -> list[dict]:
	conditions = ""
	values = {}
	
	if filters:
		if filters.get("admission_year"):
			conditions += " AND academic_year = %(admission_year)s"
			values["admission_year"] = filters.get("admission_year")
		if filters.get("campus"):
			conditions += " AND campus = %(campus)s"
			values["campus"] = filters.get("campus")
		if filters.get("program"):
			conditions += " AND program = %(program)s"
			values["program"] = filters.get("program")
		if filters.get("payment_mode"):
			conditions += " AND payment_mode = %(payment_mode)s"
			values["payment_mode"] = filters.get("payment_mode")
		if filters.get("from_date"):
			conditions += " AND payment_date >= %(from_date)s"
			values["from_date"] = filters.get("from_date")
		if filters.get("to_date"):
			conditions += " AND payment_date <= %(to_date)s"
			values["to_date"] = filters.get("to_date")

	data = frappe.db.sql(f"""
		SELECT 
			name, applicant, applicant_name, program, academic_year,
			campus, payment_date, payment_mode, transaction_id, 
			total_amount, currency
		FROM `tabApplicant Payment Receipt`
		WHERE docstatus = 1 {conditions}
		ORDER BY payment_date DESC, name DESC
	""", values, as_dict=1)
	
	return data

File: slcm/admission/report/applicant_fee_payments/applicant_fee_payments.js
// Copyright (c) 2026, TFSS and contributors
// For license information, please see license.txt

frappe.query_reports["Applicant Fee Payments"] = {
    "filters": [
        {
            "fieldname": "admission_year",
            "label": __("Admission Year"),
            "fieldtype": "Link",
            "options": "Admission Year",
            "default": ""
        },
        {
            "fieldname": "campus",
            "label": __("Campus"),
            "fieldtype": "Link",
            "options": "Campus"
        },
        {
            "fieldname": "program",
            "label": __("Programme"),
            "fieldtype": "Link",
            "options": "Programme"
        },
        {
            "fieldname": "payment_mode",
            "label": __("Payment Mode"),
            "fieldtype": "Select",
            "options": "\nOnline\nCash\nCheque\nNeft/Rtgs"
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1)
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today()
        }
    ],
    "onload": function (report) {
        report.page.add_inner_button(__("Refresh"), function () {
            report.refresh();
        });
        report.chart_options = { height: 300 };
        report.refresh();
    },
    "get_chart_data": function (columns, result) {
        if (!result || result.length === 0) return null;

        let mode_counts = {};
        result.forEach(d => {
            let mode = d.payment_mode || __("Not Specified");
            mode_counts[mode] = (mode_counts[mode] || 0) + 1;
        });

        let labels = Object.keys(mode_counts);
        let values = Object.values(mode_counts);

        return {
            data: {
                labels: labels,
                datasets: [{
                    name: __("Transactions"),
                    values: values
                }]
            },
            type: 'donut',
            height: 300,
            colors: ['#42a5f5', '#66bb6a', '#ffa726', '#ef5350', '#ab47bc', '#8d6e63', '#78909c']
        };
    },
    "get_report_summary": function (columns, result) {
        if (!result || result.length === 0) return [];

        let total_amount = 0;
        result.forEach(d => {
            total_amount += flt(d.total_amount);
        });

        return [
            {
                value: result.length,
                indicator: "Blue",
                label: __("Total Transactions"),
                datatype: "Int",
            },
            {
                value: total_amount,
                indicator: "Green",
                label: __("Total Amount Paid"),
                datatype: "Currency",
            }
        ];
    }
};
